design-proposals/dynamic-primary-pod-nic-name.md

```diff
@@ -0,0 +1,481 @@
+
+Currently, the `VirtualMachineInstanceStatus` struct holds a slice of interface status objects:
+
+```go
+type VirtualMachineInstanceStatus struct {
+...
+Interfaces []VirtualMachineInstanceNetworkInterface `json:"interfaces,omitempty"`
+...
+}
+```
+
+A new field will be added to the `VirtualMachineInstanceNetworkInterface` struct which will be used to report
+the pod interface name connected to the network the VM interface should connect to:
+
+```go
+type VirtualMachineInstanceNetworkInterface struct {
+...
+// PodInterfaceName represents the name of the pod network interface
+PodInterfaceName string `json:"podInterfaceName,omitempty"`
+...
+}
+```
+
+The VMI controller will set this field's value from the following sources:
+
+- The output of the existing naming scheme calculation based on VMI.Spec
+- Multus' network-status annotation
+
+virt-handler and virt-launcher will be adjusted to read this value instead of performing an independent naming scheme calculation.
+
+For newly created VMs - the VMI object will be handed over to the virt-handler after the `PodInterfaceName` fields will be filled.
+virt-handler in turn - will send virt-launcher the `SyncVMI` command with all the `PodInterfaceName` data present.
+
+For existing VMs, the information required for network setup is already saved in the persistent cache.
+
+In case the `PodInterfaceName` field will be empty, virt-handler and virt-launcher will fall back to the existing naming scheme calculation based on the VMI.spec.
+
+#### API Example
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
+metadata:
+  name: myvmi
+spec:
+  ...
+status:
+  ...
+  interfaces:
+      - infoSource: 'domain, guest-agent, multus-status'
+        podInterfaceName: custom-iface
+        interfaceName: eth0
+        ipAddress: 10.131.0.29
+        ipAddresses:
+          - 10.131.0.29
+        mac: '52:54:00:c0:35:78'
+        name: passtnet
+        queueCount: 1
+  ...
+```
+
+#### Pros
+
+- Manual configuration is not required
+- Provides backward compatibility
+- This solution has the potential to ease future changes in the name generation, as it is generated once at the
+  controller and not at each component (which may cause compatibility issues when the system is upgraded).
+
+#### Cons
+
+- Adds API
+- Exposes an implementation detail to VM owners which is probably not useful for them.
+
+### Alternative 2 - Report Primary Pod Interface Name Per VM
+
+Currently, the `VirtualMachineInstanceStatus` struct holds a slice of interface status objects:
+
+```go
+type VirtualMachineInstanceStatus struct {
+...
+Interfaces []VirtualMachineInstanceNetworkInterface `json:"interfaces,omitempty"`
+...
+}
+```
+
+A new field will be added to the `VirtualMachineInstanceStatus` struct, which will be used to report the pod's primary
+interface's name:
+
+```Go
+type VirtualMachineInstanceStatus struct {
+...
+Interfaces []VirtualMachineInstanceNetworkInterface `json:"interfaces,omitempty"`
```

Comment:
@orelmisan I think this is confusing to mention the vmi.Status.[]VirtualMachineInstanceNetworkInterface
when this alternative is about adding a new field to vmi.Status

Reply:
It was removed.